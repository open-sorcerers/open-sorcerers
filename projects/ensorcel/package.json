{
  "name": "ensorcel",
  "version": "0.0.2-beta.2",
  "description": "common tools for sorcerers",
  "main": "ensorcel.js",
  "module": "ensorcel.mjs",
  "repository": "//github.com/open-sorcerers/open-sorcerers",
  "author": "brekk",
  "license": "MIT",
  "private": false,
  "dependencies": {
    "fluture": "^12.1.1",
    "ramda": "^0.26.1"
  },
  "devDependencies": {
    "@rollup/plugin-buble": "^0.21.1",
    "@rollup/plugin-json": "^4.0.2",
    "@rollup/plugin-node-resolve": "^7.1.0",
    "babel-eslint": "^10.0.3",
    "eslint": "^6.8.0",
    "eslint-config-prettier": "^6.10.0",
    "eslint-config-sorcerers": "^0.0.2",
    "eslint-config-standard": "^14.1.0",
    "eslint-plugin-jsdoc": "^21.0.0",
    "eslint-plugin-node": "^11.0.0",
    "eslint-plugin-prettier": "^3.1.2",
    "eslint-plugin-standard": "^4.0.1",
    "eslint-plugin-unicorn": "^16.0.0",
    "prettier": "^1.19.1",
    "prettier-eslint": "^9.0.1",
    "rollup": "^1.31.0",
    "rollup-plugin-progress": "^1.1.1"
  }
}
