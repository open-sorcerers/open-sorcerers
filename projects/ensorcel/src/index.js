export { version } from "../package.json"
export const name = "ensorcel"
export { uncurryN } from "ramda"
export {
  futurizeWithCancel,
  unfuturize,
  downgrade,
  FLUTURE_METHOD_ARITIES
} from "./future"
export { freeze, j2 } from "./utils"
