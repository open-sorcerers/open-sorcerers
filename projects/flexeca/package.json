{
  "name": "flexeca",
  "version": "0.0.0",
  "description": "execa x fluture",
  "main": "flexeca.js",
  "module": "flexeca.mjs",
  "repository": "//github.com/open-sorcerers/open-sorcerers",
  "author": "brekk",
  "license": "MIT",
  "private": false,
  "dependencies": {
    "ensorcel": "0.0.2",
    "execa": "^4.0.0",
    "fluture": "^12.2.0",
    "ramda": "^0.26.1"
  }
}
