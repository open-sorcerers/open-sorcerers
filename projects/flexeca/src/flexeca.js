import { Future } from "fluture"
import { either, curry, ifElse, pipe, propOr } from "ramda"
import execa from "execa"
/* import { trace } from "xtrace" */

const orFalse = propOr(false)
const then = curry((fn, P) => P.then(fn))
const snag = curry((fn, P) => P.catch(fn))

export const flexeca = curry(
  (opts, cmd, args) =>
    new Future((bad, good) => {
      let ref = null

      const canceller = () => {
        ref && ref.cancel()
      }
      pipe(
        () => execa(cmd, args, opts),
        x => {
          ref = x
          return x
        },
        then(
          pipe(
            ifElse(
              either(orFalse("stderr"), orFalse("isCancelled")),
              ifElse(orFalse("stderr"), orFalse("stderr"), canceller),
              propOr("", "stdout")
            ),
            good
          )
        ),
        snag(bad)
      )()
      return canceller
    })
)
