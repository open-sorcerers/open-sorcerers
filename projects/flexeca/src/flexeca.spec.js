import { fork as rawFork } from "fluture"
import { tacit } from "ensorcel"
import { flexeca } from "./flexeca"

const cmd = "node"
const cwd = process.cwd()
const runLocally = flexeca({ cwd }, cmd)

const fork = tacit(2, rawFork)

test("flexeca", done => {
  fork(
    done,
    e => {
      expect(e).toMatchSnapshot()
      done()
    },
    runLocally(["./src/fixture.js"])
  )
})

test("flexeca - slow", done => {
  fork(
    done,
    e => {
      expect(e).toMatchSnapshot()
      done()
    },
    runLocally(["./src/fixture2.js"])
  )
})

test("flexeca - error", done => {
  fork(
    e => {
      expect(e.shortMessage).toEqual(
        "Command failed with exit code 1: node ./src/not-a-file.js"
      )
      done()
    },
    done,
    runLocally(["./src/not-a-file.js"])
  )
})

test("flexeca - cancelled", done => {
  const cancel = fork(done, done, runLocally(["./src/fixture2.js"]))

  cancel()
  setTimeout(() => {
    done()
  }, 1)
})
