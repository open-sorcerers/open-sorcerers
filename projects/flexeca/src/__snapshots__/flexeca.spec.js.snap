// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flexeca - slow 1`] = `"slow"`;

exports[`flexeca 1`] = `"cool"`;
